File: dart/lib/src/packets/src/client/pr.dart
part of '../../packets.dart';

class PrPacket extends ClientPacket {
  /// [PrPacket] is the authentication success package.
  ///
  /// This package is part of the package sent from the server to the device.
  PrPacket() : super();

  /// [fromPacket] creates a [PrPacket] from a string package in the format of `Layrz Protocol v3`.
  static PrPacket fromPacket(String raw) {
    if (!raw.startsWith('<Pr>') || !raw.endsWith('</Pr>')) {
      throw ParseException('Invalid identification package, should be <Pr>...</Pr>');
    }

    final parts = raw.substring(4, raw.length - 5).split(';');
    if (parts.length != 2) {
      throw MalformedException('Invalid package parts, should have 2 parts');
    }

    int? receivedCrc = int.tryParse(parts[1], radix: 16);
    int? calculatedCrc = calculateCrc("${parts.sublist(0, 1).join(';')};".codeUnits);

    if (receivedCrc != calculatedCrc) {
      throw CrcException('Invalid CRC, received: $receivedCrc, calculated: $calculatedCrc');
    }

    return PrPacket();
  }

  /// [toPacket] returns the package in the format of `Layrz Protocol v3`.
  @override
  String toPacket() {
    String payload = ';';
    String crc = calculateCrc(payload.codeUnits).toRadixString(16).padLeft(4, '0').toUpperCase();
    return '<Pr>$payload$crc</Pr>';
  }

  @override
  String toString() => toPacket();

  @override
  PrPacket copyWith() {
    return PrPacket();
  }
}

File: go/packets/client/pr.go
package client

import (
	"errors"
	"fmt"
	"strconv"
	"strings"

	"github.com/goldenm-software/layrz-protocol/go/v3/internal/wire"
)

// Based on the `Layrz Protocol v2` specification. PrPacket is the syncronization packet
// sent from the device to the server
type PrPacket struct{}

// FromPacket is a method that converts a raw packet to a PrPacket
// based on the `Layrz Protocol v2` specification
//
// Returns an error if the packet is invalid
func (p *PrPacket) FromPacket(raw *string) error {
	if !strings.HasPrefix(*raw, "<Pr>") || !strings.HasSuffix(*raw, "</Pr>") {
		return errors.New("invalid package, should be <Pr>...</Pr>")
	}

	*raw = strings.TrimPrefix(*raw, "<Pr>")
	*raw = strings.TrimSuffix(*raw, "</Pr>")

	parts := strings.Split(*raw, ";")
	if len(parts) != 2 {
		return errors.New("invalid package, should contain 2 parts")
	}

	receivedCrc, err := strconv.ParseUint(parts[1], 16, 16)
	if err != nil {
		return errors.New("cannot convert CRC to integer")
	}

	calculatedCrc := wire.Calculate([]byte(";"))

	if calculatedCrc != uint16(receivedCrc) {
		return fmt.Errorf("invalid CRC, received: %04X, calculated: %04X", receivedCrc, calculatedCrc)
	}

	return nil
}

// ToPacket is a method that converts a PrPacket to a raw packet
// based on the `Layrz Protocol v2` specification
func (p *PrPacket) ToPacket() *string {
	crc := wire.Calculate([]byte(";"))
	content := fmt.Sprintf("<Pr>;%04X</Pr>", crc)
	return &content
}
